crates/jarscanner/src/lib.rs

```diff
@@ -25,54 +24,76 @@ fn ends_in_class(file_name: &str) -> bool {
     file_name.ends_with(".class")
 }
 
-fn file_name_to_class_names(file_name: &str) -> Result<Vec<String>, FileNameError> {
-    if non_anon(file_name) && not_in_meta(file_name) && ends_in_class(file_name) {
-        let class_suffix = file_name.strip_suffix(".class").ok_or_else(|| {
-            FileNameError::new(format!("Failed to strip .class suffix for {}", file_name))
-        })?;
-        let final_suffix = class_suffix.strip_suffix("$").unwrap_or(class_suffix);
-
-        let dotted = final_suffix
-            .replace(r"/$", r"/")
-            .replace("$", ".")
-            .replace(r"/", ".");
-        let replace_pkg = dotted.replace(".package", "");
-        if dotted.contains(".package") {
-            Ok(vec![dotted, replace_pkg])
+fn file_name_to_class_names(file_name_ref: &str, result: &mut BTreeSet<String>) {
+    if non_anon(file_name_ref) && not_in_meta(file_name_ref) && ends_in_class(file_name_ref) {
+        let length_of_name = file_name_ref.len();
+        let mut file_name_res = String::with_capacity(length_of_name);
+        let mut was_slash = false;
+        let mut saw_k = false;
+        let mut saw_g = false;
+        let mut idx = 0;
+        // We know this ends in .class, so we're going to need to track the index of the last 6 chars
+        let end_idx = length_of_name - 6;
+        for file_char in file_name_ref.chars() {
+            if idx == end_idx {
+                break;
+            }
+
+            match file_char {
+                '/' => {
+                    was_slash = true;
+                    file_name_res.push('.')
+                }
+                '$' => {
+                    if idx == end_idx - 1 {
+                        // Do nothing
+                    } else if !was_slash {
+                        was_slash = false;
```

Comment:
I think this is a bug.

```rust
if !was_slash
```
implies `was_slash == false` and we reset it to false.

I think we need to write a test that fails assuming my claim is true that the algo is wrong, then fix it with:
```rust
if !was_slash {
  // we know was_slash == false in this branch
  file_name_res.push('.')
}
else {
  // we previously saw /
  was_slash = false;
}
```
I think we really want to change the code to track `was_dot` and just don't emit two adjacent `.` characters. So we would treat `/` and `$` the same in that case.